src/carousel/index.ts

```diff
@@ -0,0 +1,222 @@
+import { alias, tag, utils, Tag } from '@storefront/core';
+
+@alias('carousel')
+@tag('gb-carousel', require('./index.html'), require('./index.css'))
+class Carousel {
+  refs: {
+    wrapper: HTMLDivElement,
+    track: HTMLDivElement
+  };
+  props: Carousel.Props = { items: [] };
+  state: Carousel.State = {
+    settings: {
+      speed: 0,
+      slidesToShow: 1,
+      slidesToScroll: 1
+    },
+    currentSlide: 0,
+    transitioning: false,
+    items: []
+  };
+  animationEndCallback: NodeJS.Timer | null = null;
+
+  init() {
+    this.state.settings = { ...this.state.settings, ...this.props.settings };
+  }
+
+  onMount() {
+    utils.WINDOW().addEventListener('resize', this.update);
+  }
+
+  onUpdate() {
+    this.state.items = this.cloneItems();
+  }
+
+  onUnMount() {
+    utils.WINDOW().removeEventListener('resize', this.update);
+  }
+
+  moveNext = () => {
+    this.slideHandler(this.state.currentSlide + this.state.settings.slidesToScroll);
+  }
+
+  movePrevious = () => {
+    this.slideHandler(this.state.currentSlide - this.state.settings.slidesToScroll);
+  }
+
+  cloneItems = () => {
+    const numCloned = this.state.settings.slidesToShow * 2 - 1;
+    const length = this.props.items.length;
+
+    const prior = this.props.items
+      .slice(-numCloned)
+      .map((data, index) => ({ ...data, 'data-index': index - length }));
+    const posterior = this.props.items
+      .slice(0, numCloned)
+      .map((data, index) => ({ ...data, 'data-index': index }));
+    this.props.items.forEach((data, index) => (data['data-index'] = index));
```

Comment:
Isn't this doing the same work as the above statement, adding an index to the items in the array? Wouldn't
```js
this.props.items.forEach((data, index) => (data['data-index'] = index));
return prior.concat(this.props.items).concat(this.props.items.slice(0,numCloned));
```